core/src/main/scala/org/http4s/util/StreamApp.scala

```diff
@@ -14,34 +14,35 @@ abstract class StreamApp[F[_]](implicit F: Effect[F]) {
 
   private implicit val executionContext = TrampolineExecutionContext
 
-  /** Exposed for testing, so we can check exit values before the dramatic sys.exit */
-  private[util] def mainStream(args: Array[String]): Stream[F, Nothing] = {
-    val s = for {
-      shutdownRequested <- eval(signalOf[F, Boolean](false))
-      halted            <- eval(signalOf[F, Boolean](false))
-      _ <- eval(F.delay(sys.addShutdownHook {
-        unsafeRunAsync(shutdownRequested.set(true))(_ => IO.unit)
-        halted.discrete.takeWhile(_ == false).run.runAsync(_ => IO.unit).unsafeRunSync()
-      }))
-      res <- stream(args.toList, shutdownRequested.set(true))
-        .interruptWhen(shutdownRequested)
-        .onFinalize(halted.set(true))
-        .as(())
-    } yield res
-    s.drain
-  }
+  /** Adds a shutdown hook that interrupts the stream and waits for it to finish */
+  private[util] def addShutdownHook(requestShutdown: Signal[F, Boolean], halted: Signal[IO, Boolean]): F[Unit] =
+    F.delay(sys.addShutdownHook {
+      val hook = requestShutdown.set(true).runAsync(_ => IO.unit) >> halted.discrete.takeWhile(_ == false).run
+      hook.unsafeRunSync()
```

Comment:
The whole thing? It was there before I ported it to cats.

I assume the idea is to allow the server stream to finish before the JVM dies, when a shutdown is started from somewhere outside.